src/rust-crypto/backup.ts

```diff
@@ -342,20 +350,37 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
 
                 if (!request || this.stopped || !this.activeBackupVersion) {
                     logger.log(`Backup: Ending loop for version ${this.activeBackupVersion}.`);
+                    if (!request) {
+                        // nothing more to upload
+                        this.emit(CryptoEvent.KeyBackupSessionsRemaining, 0);
+                    }
                     return;
                 }
 
-                try {
-                    await this.outgoingRequestProcessor.makeOutgoingRequest(request);
```

Comment:
I find it a bit confusing that the call to `roomKeyCounts` now happens *before* the call to `makeOutgoingRequest`. I feel like it would be simpler the other way around:

```ts
                    await this.outgoingRequestProcessor.makeOutgoingRequest(request);
                    numFailures = 0;
                    if (this.stopped) break;

                    // ... comments about performance ...
                    if (!isFirstIteration) {
                        if(remainingToUploadCount === null) {
                            const keyCount = await this.olmMachine.roomKeyCounts();
                            remainingToUploadCount = keyCount.total - keyCount.backedUp;
                        } else {
                            const keysCountInBatch = this.keysCountInBatch(request);
                            remainingToUploadCount = Math.max(remainingToUploadCount - keysCountInBatch, 0);
                        }
                        this.emit(CryptoEvent.KeyBackupSessionsRemaining, remainingToUploadCount);
                    }
```

WDYT?

Reply:
Right, done here https://github.com/matrix-org/matrix-js-sdk/pull/4015/commits/3e7fa5639f609c85cd9791e24a8c275681830c9b